rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,710 @@
+also considering the ubiqitousness of JSON itself, of which JSONSchema is written in.
+
+As a developer, the primary impact of this would be in terms of needing to utilize new helpers, and
+patterns, that become required for documenting their configuration types. This would represent a
+net-new change to creating and updating components, although as part of developing this feature, all
+existing components would be bootstrapped in terms of being migrated over.
+
+In turn, developers could equate this to something like a new CI check or lint being added that told
+them when their code did not comply, and what changes needed to be made in order to do so.
+
+Overall, the cognitive overhead of this proposal would be low, as we would rely on the compiler, and
+CI, as much as possible in order to surface errors or non-compliance and explain exactly what needed
+to be changed in order to fix the errors.
+
+### Implementation
+
+#### 1. Introduce a new trait and custom derive macro: `Configurable`
+
+This trait, and derive macro, would form the basis of validating the compliance of a configuration
+as well as walking a configuration to generate its schema. There are two primary requirements for
+generating the configuration schema: discoverability and compliance.
+
+**Discoverability**, or the actual logic of inspecting the configuration types, encoding information
+about their fields, allowable values, and so on, is the most obvious requirement. We need to be able
+to find these types, know they can be inspected, and then actually do the work. **Compliance** is
+perhaps even more important: unless all configuration types are able to be inspected, then the
+configuration schema can never be used to correctly validate a user-supplied configuration.
+
+The `Configurable` trait would form the basis for discoverability. It would provide a minimalistic
+interface that walked the type, and walked its fields, mapping closely to the traditional "Visitor"
+pattern. The trait would allow exposing common items such as name, description, allowable type,
+units, and so on. Additionally, it would allow for defining custom metadata, or extensions, that
+could be parsed by external code to satisfy more advanced workflows i.e. configuration migration,
+testing, etc.
+
+The `Configurable` derive macro would form the basis for compliance. While it would generally
+provide the scaffolding to generate an implementation of the `Configurable` trait -- walking each
+field, gathering attributes and doc comments and so on -- it would also be able to validate that
+those things exists at compile time. As an example, we can enforce that all `Configuration`
+implementors are fully documented: the type itself, their subfields, and so on. It would then become
+extremely hard for developers to add new fields or types in configuration that weren't documented
+upon cutting a new Vector release.
```

Comment:
Well, initially, half and half, most likely.  We can't really enforce the stuff that has to be done at run-time unless we run something to actually check that.  This is why the release step would become relevant, because it'd be when we tried to generate the schema that it would hit those run-time checks and flip out.  We would/could eventually extend that to run in CI.

Some things would be straight up compile-time, though, and break visibly during the development cycle.